crates/subspace-networking/src/node.rs

```diff
@@ -281,14 +332,29 @@ impl Node {
             .await?;
 
         // TODO: A wrapper that'll immediately cancel query on drop
-        Ok(result_receiver)
+        Ok(WrapperWithPermits::new(
+            result_receiver,
+            vec![kademlia_permit, global_permit],
+        ))
     }
 
     pub async fn put_value(
         &self,
         key: Multihash,
         value: Vec<u8>,
     ) -> Result<impl Stream<Item = ()>, PutValueError> {
+        let kademlia_permit = self
```

Comment:
The order of the semaphore access is important. With some combination of the permit number, we could experience race conditions. Let's have two wrapper methods for all kademlia- and non-kademlia-methods with the fixed order of acquiring resources. The wrapper method should acquire semaphores and return a `WrapperWithPermits` for the inner command.

Reply:
This will also reduce similar copy-paste code in the methods.

Reply:
I made two semaphores non-overlapping, so you only need to take one and not both. Hope it helps. And you are right about ordering, it can be a source of issues.